File: src/tangle-verification/tangle-verification.go
package main

import (
	"fmt"
	"sync"
	"tangle_node"
)

var wg sync.WaitGroup

var queue = make([]*tangle_node.Node, 0)
var queueWriteMutex = make(chan bool, 1)

func addWork() {
	defer wg.Done()
	newNode := tangle_node.New()

	var isComplete int
	for verified, i := 0, 0; verified < 2; i++ {
		var comm = make(chan bool, 1)
		var cb = make(chan bool)
		fmt.Println(len(queue))
		go queue[i%len(queue)].Verify(newNode, comm, cb)
		// ** node evaluation work would go here **
		comm <- true
		didVerify := <- cb
		if didVerify { verified++ }
	}
	<- queueWriteMutex
	queue = append(queue, newNode)
	if isComplete {
		queue = append(queue[:i%len(queue)], queue[((i%len(queue))+1):]...)
	}
	queueWriteMutex <- true
}

// Execution code
func main() {
	queue = append(queue, tangle_node.New())
	// head := queue[0]
	queue = append(queue, tangle_node.New())
	queueWriteMutex <- true

	for i:=0; i<80; i++ {
		wg.Add(1)
		go addWork()
	}

	wg.Wait()
}

File: src/tangle-verification/tangle-verification.py
import threading
import queue

class tangle_node:
    verified = False
    approvalCount = 0
    approvalMutex = queue.Queue(maxsize=1)
    workingMutex = queue.Queue(maxsize=2)
    outbound = []

    def __init__(self):
        self.approvalMutex.put(True)
        return None
    
    def Verify(self, addr, approved, cb):
        self.approvalMutex.get()
        if self.approvalCount >= 2 or addr in self.outbound:
            cb.put(False)
            self.approvalMutex.put(True)
            return
        self.approvalMutex.put(True)

        self.workingMutex.put(True)
        verified = self.approved.get()

        if verified:
            self.approvalMutex.get()
            if self.approvalCount >= 2 or addr in self.outbound:
                cb.put(False)
                self.approvalMutex.put(True)
                self.workingMutex.get()
                return
            else:
                self.approvalCount += 1
            
            if self.approvalCount == 2:
                self.verified = True

            self.outbound.append(addr)
            self.approvalMutex.put(True)

        self.workingMutex.get()
        cb.put(True)

l = []
lWriteMutex = queue.Queue(maxsize=1)

def addWord():
    newNode = tangle_node()
    
    verified, i = 0, 0
    while verified < 2:
        comm = queue.Queue(maxsize=1)
        cb = queue.Queue(maxsize=0)
        l[i%len(l)].Verify(newNode, comm, cb)
        #  ** node evaluation work would go here **
        comm.put(True)
        didVerify = cb.get()
        if didVerify:
            verified += 1
        i+= 1

if __name__ == "__main__":
    l.append(tangle_node())
    l.append(tangle_node())
    lWriteMutex.put(True)

    for i in range(2) :
        addWord()

